Mathlib/Analysis/PSeries.lean

```diff
@@ -19,134 +19,180 @@ converges if and only if so does `∑ k, 2 ^ k f (2 ^ k)`. We prove this test in
-  convert add_le_add_left (nsmul_le_nsmul_right (sum_condensed_le' hf n) 2) (f 1)
-  simp [sum_range_succ', add_comm, pow_succ, mul_nsmul', sum_nsmul]
-#align finset.sum_condensed_le Finset.sum_condensed_le
 
+theorem sum_schlomilch_le {C : ℕ} (hf : ∀ ⦃m n⦄, 1 < m → m ≤ n → f n ≤ f m) (h_pos : ∀ n, 0 < u n)
+    (h_nonneg : ∀ n, 0 ≤ f n) (hu : Monotone u) (h_succ_diff : SuccDiffBounded C u) (n : ℕ) :
+    ∑ k in range (n + 1), (u (k + 1) - u k) • f (u k) ≤
+    (u 1 - u 0) • f (u 0) + C • ∑ k in Ico (u 0 + 1) (u n + 1), f k := by
+  rw [sum_range_succ', add_comm]
+  gcongr
+  suffices ∑ k in range n, (u (k + 2) - u (k + 1)) • f (u (k + 1)) ≤
+  C • ∑ k in range n, ((u (k + 1) - u k) • f (u (k + 1))) by
+    refine this.trans (nsmul_le_nsmul_right ?_ _)
+    exact sum_schlomilch_le' hf h_pos hu n
+  have : ∀ k ∈ range n, (u (k + 2) - u (k + 1)) • f (u (k + 1)) ≤
+    C • ((u (k + 1) - u k) • f (u (k + 1))) := by
+    intro k _
+    rw [smul_smul]
+    gcongr
+    · exact h_nonneg (u (k + 1))
+    exact mod_cast h_succ_diff k
+  convert sum_le_sum this
+  simp [smul_sum]
 end Finset
 
 namespace ENNReal
 
-open Filter BigOperators
+open Filter BigOperators Finset
 
-variable {f : ℕ → ℝ≥0∞}
+variable {u : ℕ → ℕ} {f : ℕ → ℝ≥0∞}
 
-theorem le_tsum_condensed (hf : ∀ ⦃m n⦄, 0 < m → m ≤ n → f n ≤ f m) :
-    ∑' k, f k ≤ f 0 + ∑' k : ℕ, 2 ^ k * f (2 ^ k) := by
-  rw [ENNReal.tsum_eq_iSup_nat' (Nat.tendsto_pow_atTop_atTop_of_one_lt _root_.one_lt_two)]
-  refine' iSup_le fun n => (Finset.le_sum_condensed hf n).trans (add_le_add_left _ _)
-  simp only [nsmul_eq_mul, Nat.cast_pow, Nat.cast_two]
+open NNReal in
+theorem le_tsum_schlomilch (hf : ∀ ⦃m n⦄, 0 < m → m ≤ n → f n ≤ f m) (h_pos : ∀ n, 0 < u n)
+    (hu_strict : StrictMono u) :
+    ∑' k , f k ≤ ∑ k in range (u 0), f k + ∑' k : ℕ, (u (k + 1) - u k) * f (u k) := by
+  have hu : Monotone u := by
+    apply StrictMono.monotone hu_strict
+  rw [ENNReal.tsum_eq_iSup_nat' (StrictMono.tendsto_atTop hu_strict)]
+  refine' iSup_le fun n => (Finset.le_sum_schlomilch hf h_pos hu n).trans (add_le_add_left _ _)
+  have (k : ℕ) : ((u (k + 1) : ℝ≥0∞) - (u k : ℝ≥0∞) : ℝ≥0∞) = (u (k + 1) - (u k : ℕ) : ℕ) := by
+    simp [NNReal.coe_sub (Nat.cast_le (α := ℝ≥0).mpr <| (hu_strict k.lt_succ_self).le)]
+  simp only [nsmul_eq_mul, this]
   apply ENNReal.sum_le_tsum
-#align ennreal.le_tsum_condensed ENNReal.le_tsum_condensed
 
-theorem tsum_condensed_le (hf : ∀ ⦃m n⦄, 1 < m → m ≤ n → f n ≤ f m) :
-    (∑' k : ℕ, 2 ^ k * f (2 ^ k)) ≤ f 1 + 2 * ∑' k, f k := by
-  rw [ENNReal.tsum_eq_iSup_nat' (tendsto_atTop_mono Nat.le_succ tendsto_id), two_mul, ← two_nsmul]
+theorem tsum_schlomilch_le {C : ℕ} (hf : ∀ ⦃m n⦄, 1 < m → m ≤ n → f n ≤ f m) (h_pos : ∀ n, 0 < u n)
+    (h_nonneg : ∀ n, 0 <= f n) (hu_strict : StrictMono u) (h_succ_diff : SuccDiffBounded C u) :
+    ∑' k : ℕ, (u (k + 1) - u k) * f (u k) ≤ (u 1 - u 0) * f (u 0) + C * ∑' k, f k := by
+  have hu : Monotone u := by
```

Comment:
You can simplify this a bit further:
- an `apply` on the last line usually be converted to `exact`, an usually this is desirable
(There are exceptions to this rule, but "always use exact" is a good first rule of thumb.)
- `by exact` can be omitted --- then you just give a proof term, no need to use tactic mode
- you can use dot notation: since `hu_strict` has type `StrictMono` and you're applying a lemma from that namespace, you can also write this as `hu_strict.monotone`.
Altogether, you obtain `have hu : Monotone u := hu_strict.monotone`.

- at this point, you can consider "inlining" this `have`, i.e. replacing it by its proof whenever you use it.
My rule of thumb is that if a proof term is long or you use it at least three times, you should probably extract it. If it's short and you use it once (like here), I would inline.